File: scripts/corpus.py
from dataclasses import dataclass
from dataclasses import field
from enum import StrEnum
from pathlib import Path
from xml.etree import cElementTree as ET


class ResourceType(StrEnum):
    """

    ResourceType is an enumerated type representing different types of resource identifiers in the .NET environment.

    Attributes:
        NET_STANDARD_ASSEMBLY (str): Represents a .NET standard assembly.
        ASSEMBLY (str): Represents a general assembly.
        DLL (str): Represents a DLL (Dynamic Link Library) file.
    """

    NET_STANDARD_ASSEMBLY = "NetStandardAssembly".casefold()
    ASSEMBLY = "Assembly".casefold()
    DLL = "Dll".casefold()


@dataclass(slots=True)
class Resource:
    """
    Resource class representing a resource entity.

    Attributes:
        type (ResourceType): The type of the resource.
        name (str): The name of the resource.
        optional (bool): Flag indicating whether the resource is optional. Defaults to False.
        root (str | None): Root directory path for the resource. Defaults to None.
    """

    type: ResourceType
    name: str
    optional: bool = field(default=False)
    root: str | None = field(default=None)


@dataclass(slots=True)
class ResourceBundle:
    """
    ResourceBundle represents a collection of resources managed as a bundle.

    Attributes:
        root (Path): The root directory path containing the resources.
        resources (list[Resource]): A list of resource objects included in the bundle.
        versioned (bool, optional): Indicates if the resource bundle is versioned. Defaults to False.
    """

    root: Path
    resources: list[Resource]
    versioned: bool = field(default=False)


@dataclass(slots=True)
class Corpus:
    """
    Represents a collection of resource bundles.

    Attributes:
        bundles (list[ResourceBundle]): A list containing instances of ResourceBundle.

    """

    bundles: list[ResourceBundle]


def load_corpus(path: Path) -> Corpus:
    """
    Parses an XML file located at the given path and returns a Corpus object. The XML file should contain resource bundles within a 'Resources' root element.

    Args:
        path: Path to the XML file that contains the corpus data.

    Returns:
        A Corpus object containing parsed resource bundles and their corresponding resources.

    Raises:
        ValueError: If the XML file is malformed, specifically if the 'Resources' element is missing within the 'Corpus' element.
    """
    with path.open("r", encoding="utf-8") as file:
        xml = ET.ElementTree(file=file)

    resources_root = xml.getroot().find("Resources")

    if resources_root is None:
        raise ValueError(
            "Corpus file is malformed, and should contain a root 'Resources' element within the 'Corpus' element"
        )

    bundles: list[ResourceBundle] = []
    for bundle in resources_root:
        bundle_root = Path(bundle.attrib["Root"])
        bundle_versioned = (
            True
            if bundle.attrib.get("Versioned", "false").casefold() == "true".casefold()
            else False
        )

        bundle_obj = ResourceBundle(bundle_root, [], bundle_versioned)
        bundles.append(bundle_obj)

        for resource in bundle:
            resource_type = ResourceType(resource.attrib["Type"].casefold())
            resource_name = resource.attrib["Name"]
            resource_optional = (
                True
                if resource.attrib.get("Optional", "false").casefold()
                == "true".casefold()
                else False
            )
            resource_root = resource.attrib.get("Root", None)

            bundle_obj.resources.append(
                Resource(resource_type, resource_name, resource_optional, resource_root)
            )

    return Corpus(bundles)

File: src/ColonySync.Mod.Bootloader/Corpus.cs
// MIT License
//
// Copyright (c) 2024 SirRandoo
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

using System;
using System.Xml.Serialization;
using NetEscapades.EnumGenerators;

namespace ColonySync.Mod.Bootloader;

/// <summary>Represents a collection of resource bundles in the mod's bootloader system.</summary>
public class Corpus(ResourceBundle[] resources)
{
    /// <summary>
    ///     Contains an array of ResourceBundle objects, each representing a bundle of resources that
    ///     can be loaded and processed by the application.
    /// </summary>
    [XmlArrayItem(typeof(ResourceBundle))]
    public ResourceBundle[] Resources { get; } = resources ?? throw new ArgumentNullException(nameof(resources));
}

/// <summary>
///     Represents a collection of resources within a specified root directory, optionally
///     versioned, and encompassing multiple individual resources.
/// </summary>
public class ResourceBundle(string root, bool versioned, Resource[] resources)
{
    /// <summary>
    ///     Specifies the root directory for the resource bundle. This property holds the base path
    ///     from where the resources of the bundle will be located or processed.
    /// </summary>
    [XmlAttribute]
    public string Root { get; } = root ?? throw new ArgumentNullException(nameof(root));

    /// <summary>
    ///     Indicates whether the resource bundle supports versioning, influencing how the resources
    ///     are managed and loaded.
    /// </summary>
    [XmlAttribute]
    public bool Versioned { get; } = versioned;

    /// <summary>
    ///     Contains the list of resources defined in the resource bundle, which can be processed and
    ///     loaded by the application.
    /// </summary>
    [XmlElement(elementName: "Resource", typeof(Resource))]
    public Resource[] Resources { get; } = resources ?? throw new ArgumentNullException(nameof(resources));
}

/// <summary>Represents a file on disk.</summary>
public class Resource(ResourceType type, string name, bool optional, string root)
{
    /// <summary>
    ///     Specifies the type of the resource, defining how it should be processed and loaded by the
    ///     application.
    /// </summary>
    [XmlAttribute]
    public ResourceType Type { get; init; } = type;

    /// <summary>The name of the resource file.</summary>
    [XmlAttribute]
    public string Name { get; init; } = name ?? throw new ArgumentNullException(nameof(name));

    /// <summary>Indicates whether the resource is optional.</summary>
    [XmlAttribute]
    public bool Optional { get; init; } = optional;

    /// <summary>
    ///     Specifies the root directory for the resource. This path is typically combined with the
    ///     overall path to correctly locate the resource on disk.
    /// </summary>
    [XmlAttribute]
    public string Root { get; init; } = root ?? throw new ArgumentNullException(nameof(root));
}

/// <summary>Specifies different types of resources that the mod's bootloader can load.</summary>
[EnumExtensions]
public enum ResourceType
{
    /// <summary>Represents a dynamic-link library (DLL) resource type.</summary>
    [XmlEnum(Name = "Dll")] Dll,

    /// <summary>Represents an assembly that can be loaded by the mod's bootloader.</summary>
    [XmlEnum(Name = "Assembly")] Assembly,

    /// <summary>Represents a .NET Standard assembly, which is compatible with multiple .NET platforms.</summary>
    [XmlEnum(Name = "NetStandardAssembly")]
    NetStandardAssembly
}